apiserver/facades/client/usermanager/usermanager.go

```diff
@@ -252,47 +313,42 @@ func (api *UserManagerAPI) UserInfo(ctx context.Context, request params.UserInfo
 		}
 	}
 
-	var infoForUser = func(user *state.User) params.UserInfoResult {
-		var lastLogin *time.Time
-		userLastLogin, err := user.LastLogin()
-		if err != nil {
-			if !state.IsNeverLoggedInError(err) {
-				api.logger.Debugf("error getting last login: %v", err)
-			}
-		} else {
-			lastLogin = &userLastLogin
-		}
+	var infoForUser = func(tag names.UserTag, user coreuser.User) params.UserInfoResult {
 		result := params.UserInfoResult{
 			Result: &params.UserInfo{
-				Username:       user.Name(),
-				DisplayName:    user.DisplayName(),
-				CreatedBy:      user.CreatedBy(),
-				DateCreated:    user.DateCreated(),
-				LastConnection: lastLogin,
-				Disabled:       user.IsDisabled(),
+				Username:       user.Name,
+				DisplayName:    user.DisplayName,
+				CreatedBy:      user.CreatorUUID.String(),
+				DateCreated:    user.CreatedAt,
+				LastConnection: &user.LastLogin,
+				Disabled:       user.Disabled,
 			},
 		}
-		if user.IsDisabled() {
+		if user.Disabled {
 			// disabled users have no access to the controller.
 			result.Result.Access = string(permission.NoAccess)
 		} else {
-			accessForUser(user.UserTag(), &result)
+			accessForUser(tag, &result)
 		}
 		return result
 	}
 
 	argCount := len(request.Entities)
 	if argCount == 0 {
-		users, err := api.state.AllUsers(request.IncludeDisabled)
+
+		// Get all users
+		users, err := api.userService.GetAllUsers(ctx)
```

Comment:
I commented on this when I reviewed it last time but my comment wasn't saved :-(
This block of code is business logic which does not belong here. We want all users if the admin is doing the operation, or just the logged in user otherwise. This is better implemented by doing this (pseudo code)
```
var f *Filter
if !isAdmin {
    f = &Filter{UserName: apiUser.Name()}
}
users, err := api.userService.GetUsers(f)
```
We load exactly the users we want by using a filtered query on the user manager service. The filter is set up according to whether we are the admin or not.

Then below all we need is
```
for _, user := range users {
	userTag := names.NewLocalUserTag(user.Name)
	results.Results = append(results.Results, infoForUser(userTag, user))
}
```
We don't need the filtering business logic here any more.

Reply:
Done. Thx for advice.